### DISCUSSION TITLE
Is it possible to add a `prefix` or `url_prefix` argument to SimpleRouter or BaseRouter like Flask Blueprint's `url_prefix` argument

### DISCUSSION THREAD

**Discussion by ryomahan** [encode/django-rest-framework#9575] [Ideas & Suggestions] [answered]
For example, I create a user app and define user's url.py like this:

```python
from rest_framework import routers

router = routers.SimpleRouter()
router.register(r'users', UserViewSet)
router.register(r'accounts', AccountViewSet)
urlpatterns = router.urls
```

now I want add version info to url path, I need add version info to every register function, like this:

```python
from rest_framework import routers

router = routers.SimpleRouter()
router.register(r'v1/users', UserViewSet)
router.register(r'v1/accounts', AccountViewSet)
urlpatterns = router.urls
```

I don't find a simple way to do it, I remembered that Flask can do this by define Blueprint's `url_prefix` arguments, so Is it possible to add a `prefix` or `url_prefix` argument to SimpleRouter or BaseRouter like Flask Blueprint's `url_prefix` argument? like this:

 ```python
from rest_framework import routers

router = routers.SimpleRouter(prefix="v1")
router.register(r'users', UserViewSet)
router.register(r'accounts', AccountViewSet)
urlpatterns = router.urls
```

I tried to update `rest_framework.routers.SimpleRouter` like this:

```python
class SimpleRouter(BaseRouter):
    def __init__(self, prefix="", trailing_slash=True, use_regex_path=True):
        self.prefix = prefix
        ...

    def get_urls(self):
        ...
        for prefix, viewset, basename in self.registry:
            prefix = f"{self.prefix}/{prefix}" if self.prefix else prefix
        ...
```

it works, but maybe no strong enough. Is that a good suggestion? If so I'd like to continue to refine it.

**Comment by browniebroke** [comment]
I would receommend putting the v1 and v2 routers each in their own folder with a `urls` module:

```python
# apps/api/v1/urls.py
from rest_framework import routers

router = routers.SimpleRouter()
router.register(r"users", UserViewSetV1)
urlpatterns = router.urls
```

```python
# apps/api/v2/urls.py
from rest_framework import routers

router = routers.SimpleRouter()
router.register(r"users", UserViewSetV2)
urlpatterns = router.urls
```

Then you can have a `urls` module at the top of you `api` app and use the [standard Django's `include`](https://docs.djangoproject.com/en/5.1/ref/urls/#include) to prefix all routes as follow:

```python
# apps/api/urls.py
from django.urls import include, path
from rest_framework.authtoken import views

urlpatterns = [
    path("v1/", include("apps.api.v1.urls")),
    path("v2/", include("apps.api.v2.urls")),
]
```

Hope that helps

**Reply by ryomahan** [reply]
yeah, that's also a good way to resolve my question.

**Reply by browniebroke** [reply, marked as answer]
To answer your question:

> Is it possible to add a `prefix` or `url_prefix` argument to SimpleRouter or BaseRouter like Flask Blueprint's `url_prefix` argument?

I think the answer will be no. I'm not a maintainer, but DRF is unlikely to accept this kind of requests, as there is a good way of handling this case:

> At this point in its lifespan we consider Django REST framework to be feature-complete. We focus on pull requests that track the continued development of Django versions, and generally do not accept new features or code formatting changes.

https://www.django-rest-framework.org/community/contributing/

**Reply by ryomahan** [reply]
Thks for the info.
